app/lib/navigation/app_router.dart

```diff
@@ -0,0 +1,191 @@
+import 'package:flutter_bloc/flutter_bloc.dart';
+import 'package:logging/logging.dart';
+import 'package:prototype/conversation_pane/conversation_details/add_members.dart';
+import 'package:prototype/conversation_pane/conversation_details/conversation_details.dart';
+import 'package:prototype/conversation_pane/conversation_pane.dart';
+import 'package:prototype/homescreen.dart';
+import 'package:prototype/intro_screen.dart';
+import 'package:prototype/registration/registration.dart';
+import 'package:prototype/settings/developer.dart';
+import 'package:prototype/settings/user.dart';
+import 'package:prototype/theme/theme.dart';
+
+import 'navigation_cubit.dart';
+
+final _log = Logger('AppRouter');
+
+class EmptyConfig {
+  const EmptyConfig();
+}
+
+class AppRouter implements RouterConfig<EmptyConfig> {
+  AppRouter();
+
+  final AppRouterDelegate _routerDelegate = AppRouterDelegate();
+
+  final AppBackButtonDispatcher _backButtonDispatcher =
+      AppBackButtonDispatcher();
+
+  @override
+  BackButtonDispatcher? get backButtonDispatcher => _backButtonDispatcher;
+
+  @override
+  RouteInformationParser<EmptyConfig>? get routeInformationParser => null;
+
+  @override
+  RouteInformationProvider? get routeInformationProvider => null;
+
+  @override
+  RouterDelegate<EmptyConfig> get routerDelegate => _routerDelegate;
+}
+
+/// The main application router
+///
+/// Builds pages from the navigation state [NavigationState] provided by the
+/// [NavigationCubit]. This is where the translation from the navigation
+/// state to the actual list of pages happens.
+class AppRouterDelegate extends RouterDelegate<EmptyConfig> {
+  AppRouterDelegate();
+
+  final GlobalKey<NavigatorState> _navigatorKey = GlobalKey<NavigatorState>();
+
+  final PageStorageBucket _bucket = PageStorageBucket();
+
+  @override
+  Widget build(BuildContext context) {
+    final navigationState = context.watch<NavigationCubit>().state;
+
+    // hide material banners if any
+    ScaffoldMessenger.of(context).hideCurrentMaterialBanner();
+
+    final screenType = context.responsiveScreenType;
+
+    // routing
+    final List<MaterialPage> pages = switch (navigationState) {
+      IntroNavigation(screens: final screensStack) => [
+          for (final screenType in screensStack)
+            MaterialPage(key: screenType.key, child: screenType.screen),
+        ],
+      HomeNavigation home => home.pages(screenType),
+    };
+
+    _log.finer(
+        "AppRouterDelegate.build: navigationState = $navigationState, pages=$pages");
+
+    return PageStorage(
+      bucket: _bucket,
+      child: Navigator(
+        key: _navigatorKey,
+        pages: pages,
+        // Note: onPopPage is deprecated, and instead we should use onDidRemovePage. However, the latter
+        // does not allow to distinguish whether the page was popped by the user or programmatically.
+        //
+        // Also see <https://github.com/flutter/flutter/issues/153122>
```

Comment:
Please create an issue if this is something we should revisit later

Reply:
Good Idea: https://github.com/phnx-im/infra/issues/244